water/lib/Dialect/Wave/IR/WaveOps.cpp

```diff
@@ -310,14 +336,67 @@ llvm::LogicalResult wave::IterateOp::verifyRegions() {
        llvm::enumerate(resultTypes, terminatorOperandTypes, iterArgTypes,
                        blockIterArgTypes)) {
     auto istr = std::to_string(i);
-    if (llvm::failed(detail::verifyTypesCompatible(
-            llvm::cast<wave::WaveTensorType>(result),
-            llvm::cast<wave::WaveTensorType>(terminatorOperand),
-            /*includeAddressSpace=*/true, getLoc(), "result #" + istr,
-            "terminator operand #" + istr))) {
-      return llvm::failure();
+
+    // Verify result type vs terminator operand type.
+    if (isa<wave::WaveTensorType>(result) &&
+        isa<wave::WaveTensorType>(terminatorOperand)) {
+      if (llvm::failed(detail::verifyTypesCompatible(
+              llvm::cast<wave::WaveTensorType>(result),
+              llvm::cast<wave::WaveTensorType>(terminatorOperand),
```

Comment:
`isa` followed by `cast` is an anti-pattern in LLVM. Just do `dyn_cast` before the if/else block similarly to what you do above.